package com.wagglex2.waggle.domain.assignment.controller;

import com.wagglex2.waggle.common.response.ApiResponse;
import com.wagglex2.waggle.common.security.CustomUserDetails;
import com.wagglex2.waggle.domain.assignment.dto.request.AssignmentCreationRequestDto;
import com.wagglex2.waggle.domain.assignment.dto.request.AssignmentUpdateRequestDto;
import com.wagglex2.waggle.domain.assignment.dto.response.AssignmentResponseDto;
import com.wagglex2.waggle.domain.assignment.service.AssignmentService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/assignments")
@RequiredArgsConstructor
public class AssignmentController {
    private final AssignmentService assignmentService;

    @PostMapping("/")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<ApiResponse<Long>> createAssignment(
            @RequestBody @Valid AssignmentCreationRequestDto requestDto,
            @AuthenticationPrincipal CustomUserDetails userDetails
    ) {
        Long assignmentId = assignmentService.createAssignment(requestDto, userDetails.getUserId());

        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok("과제 공고를 성공적으로 등록하였습니다.", assignmentId));
    }

    @GetMapping("/{assignmentId}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<ApiResponse<AssignmentResponseDto>> getAssignment(@PathVariable Long assignmentId) {
        AssignmentResponseDto responseDto = assignmentService.getAssignment(assignmentId);

        return ResponseEntity.ok(
                ApiResponse.ok("과제 공고를 성공적으로 조회하였습니다.", responseDto)
        );
    }

    @PutMapping("/{assignmentId}")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<ApiResponse<Void>> updateAssignment(
            @PathVariable Long assignmentId,
            @RequestBody @Valid AssignmentUpdateRequestDto requestDto,
            @AuthenticationPrincipal CustomUserDetails userDetails
    ) {
        assignmentService.updateAssignment(userDetails.getUserId(), assignmentId, requestDto);

        return ResponseEntity.ok(
                ApiResponse.ok("과제 공고를 성공적으로 수정하였습니다.")
        );
    }
}
